fdk/headers.py

```diff
@@ -15,29 +15,59 @@
 from fdk import constants
 
 
-def decap_headers(hdsr):
+def decap_headers(hdsr, merge=True):
     ctx_headers = {}
     if hdsr is not None:
         for k, v in hdsr.items():
             k = k.lower()
             if k.startswith(constants.FN_HTTP_PREFIX):
-                ctx_headers[k.lstrip(constants.FN_HTTP_PREFIX)] = v
-            else:
-                ctx_headers[k] = v
+                push_header(ctx_headers, k[len(constants.FN_HTTP_PREFIX):], v)
+            elif merge:
+                # http headers override functions headers in context
+                # this is not ideal but it's the more correct view from the
+                # consumer perspective than random choice and for things
+                # like host headers
+                if k not in ctx_headers:
+                    ctx_headers[k] = v
     return ctx_headers
 
 
+def push_header(input_map, key, value):
+    if key not in input_map:
+        input_map[key] = value
+        return
+
+    current_val = input_map[key]
+
+    if isinstance(current_val, list):
+        if isinstance(value, list):  # both lists concat
+            input_map[key] = current_val + value
+        else:  # copy and append current value
+            new_val = current_val.copy()
+            new_val.append(value)
+            input_map[key] = new_val
+    else:
+        if isinstance(value, list):  # copy new list value and prepend current
+            new_value = value.copy()
+            new_value.insert(0, current_val)
+            input_map[key] = new_value
+        else:  # both non-lists create a new list
+            input_map[key] = [current_val, value]
+
+
 def encap_headers(headers, status=None):
     new_headers = {}
     if headers is not None:
         for k, v in headers.items():
             k = k.lower()
+            print("looking at %s, %s-> %s" % (k, v, new_headers))
```

Comment:
nit: i guess this is a leftover from the a debug session

Reply:
ooh yeah 
thanks